wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubPersonalAccessTokenFetcher.java

```diff
@@ -0,0 +1,205 @@
+              NameGenerator.generate("oauth2-", 5),
+              NameGenerator.generate("id-", 5),
+              oAuthToken.getToken());
+      Optional<Boolean> valid = isValid(token);
+      if (valid.isEmpty()) {
+        throw new ScmCommunicationException(
+            "Unable to verify if current token is a valid GitHub token.  Token's scm-url needs to be "
+                + GithubApiClient.GITHUB_SERVER);
+      } else if (!valid.get()) {
+        throw new ScmCommunicationException(
+            "Current token doesn't have the necessary privileges. Please make sure Che app scopes are correct and containing at least: "
+                + DEFAULT_TOKEN_SCOPES.toString());
+      }
+      return token;
+    } catch (UnauthorizedException e) {
+      throw new ScmUnauthorizedException(
+          cheSubject.getUserName()
+              + " is not authorized in "
+              + OAUTH_PROVIDER_NAME
+              + " OAuth provider.",
+          OAUTH_PROVIDER_NAME,
+          "2.0",
+          getLocalAuthenticateUrl());
+    } catch (NotFoundException
+        | ServerException
+        | ForbiddenException
+        | BadRequestException
+        | ScmItemNotFoundException
+        | ScmBadRequestException
+        | ConflictException e) {
+      LOG.error(e.getMessage(), e);
+      throw new ScmCommunicationException(e.getMessage(), e);
+    }
+  }
+
+  @Override
+  public Optional<Boolean> isValid(PersonalAccessToken personalAccessToken)
+      throws ScmCommunicationException, ScmUnauthorizedException {
+    if (!githubApiClient.isConnected(personalAccessToken.getScmProviderUrl())) {
+      LOG.debug("not a valid url {} for current fetcher ", personalAccessToken.getScmProviderUrl());
+      return Optional.empty();
+    }
+
+    try {
+      String[] scopes = githubApiClient.getTokenScopes(personalAccessToken.getToken());
+      return Optional.of(Boolean.valueOf(containsScopes(DEFAULT_TOKEN_SCOPES, scopes)));
+    } catch (ScmItemNotFoundException | ScmCommunicationException | ScmBadRequestException e) {
+      LOG.error(e.getMessage(), e);
+      throw new ScmCommunicationException(e.getMessage(), e);
+    }
+  }
+
+  boolean containsScopes(Set<String> requiredScopes, String[] tokenScopes) {
+    Arrays.sort(tokenScopes);
+    // We need check that the token has the required minimal scopes.  The scopes can be normalized
+    // by GitHub, so we need to be careful for sub-scopes being included in parent scopes.
+    for (String requiredScope : requiredScopes) {
+      String parentScope = SCOPE_MAP.get(requiredScope);
+      if (parentScope == null) {
+        // requiredScope is not recognized as a GitHub scope, so just skip it.
+        continue;
+      }
+      if (Arrays.binarySearch(tokenScopes, parentScope) < 0
+          && Arrays.binarySearch(tokenScopes, requiredScope) < 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private String getLocalAuthenticateUrl() {
```

Comment:
Yes, this URL is passed to the client when OAuth login tourine should be started from scratch